chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,275 @@
+extern crate lru;
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use std::borrow::Borrow;
+use crate::Signature;
+use crate::hash_to_g2;
+use crate::aggregate_verify as agg_ver;
+use crate::gtelement::GTElement;
+use crate::PublicKey;
+use std::collections::HashMap;
+use sha2::{Digest, Sha256};
+
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyList, PyBool};
+
+pub type Bytes32 = [u8; 32];
+pub type Bytes48 = [u8; 48];
+
+#[cfg_attr(
+    feature = "py-bindings",
+    pyclass(name = "BLSCache"),
+)]
+pub struct BLSCache {
+    cache: LruCache<Bytes32, GTElement>,
+}
+
+impl BLSCache {
+    
+    pub fn generator(cache_size: Option<usize>) -> Self {
+        let cache: LruCache<Bytes32, GTElement> = LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self{cache}
+    }
+    
+    pub fn get_pairings<P: Borrow<[Bytes48]>, M: Borrow<[Vec<u8>]>> (
+        &mut self,
+        pks: &P,
+        msgs: &M,
+        force_cache: bool,
+    ) -> Vec<GTElement> {
+        let mut pairings: Vec<Option<GTElement>> = vec![];
+        let mut missing_count: usize = 0;
+        
+        for (pk, msg) in pks.borrow().iter().zip(msgs.borrow().iter()) {
+            let mut aug_msg = pk.to_vec();
+            aug_msg.extend_from_slice(msg.borrow()); // pk + msg
+            let mut hasher = Sha256::new();
+            hasher.update(aug_msg);
```

Comment:
You can call `hasher.update` multiple times, which means you don't actually have to create `aug_msg`.

```diff
-            let mut aug_msg = pk.to_vec();
-            aug_msg.extend_from_slice(msg.borrow()); // pk + msg
             let mut hasher = Sha256::new();
+            hasher.update(pk);
+            hasher.update(msg);
-            hasher.update(aug_msg);
```

Reply:
(also not tested, so some types may be off by a `&` or something)